File: examples/telegraf-bot/src/bot.ts
/**
 * A small, realistic Telegraf bot with FlowCastle attached.
 *
 * Everything below /start is ordinary Telegraf code. FlowCastle is installed once
 * with `bot.use(fc)` and then used from handlers via `ctx.flowcastle` — identify a
 * contact, record goals, hand a chat to a human, or start a flow that teammates
 * built visually in the FlowCastle editor.
 */
import { Context, Markup, Telegraf } from 'telegraf';
import { flowcastle, FlowCastleFlavor } from '@flowcastle/telegraf';

import {
  completedQualification,
  InMemoryLeadQualificationStore,
  nextQualificationStep,
  parseQualificationCallback,
  QUALIFICATION_OPTIONS,
  QUALIFICATION_QUESTIONS,
  qualificationLabel,
  saveQualificationAnswer,
  type QualificationStep,
} from './lead-qualification';

type BotContext = FlowCastleFlavor<Context>;

function requireEnv(name: string): string {
  const value = process.env[name];
  if (!value) {
    console.error(`Missing required environment variable ${name}.`);
    process.exit(1);
  }
  return value;
}

const BOT_TOKEN = requireEnv('BOT_TOKEN');
const FLOWCASTLE_API_KEY = requireEnv('FLOWCASTLE_API_KEY');
const FOLLOW_UP_FLOW_KEY = process.env.FLOWCASTLE_FOLLOW_UP_FLOW_KEY;

const bot = new Telegraf<BotContext>(BOT_TOKEN);

const fc = flowcastle<BotContext>({
  apiKey: FLOWCASTLE_API_KEY,
  // Routing-only content: commands and callback ids are shared, free text is not.
  privacy: { contactFields: ['username', 'languageCode'] },
  // Lets flows built in the FlowCastle editor run through this bot process.
  runtime: { enabled: true },
  onError: (error) => console.error('[flowcastle]', error),
});
// Observe outgoing calls made outside middleware too (e.g. bot.telegram.sendMessage on a timer).
fc.wrapTelegram(bot.telegram);
bot.use(fc);

const qualification = new InMemoryLeadQualificationStore();

const keyboardFor = (step: QualificationStep) =>
  Markup.inlineKeyboard(
    QUALIFICATION_OPTIONS[step].map((option) => [Markup.button.callback(option.label, `lead:${step}:${option.value}`)]),
  );

bot.start(async (ctx) => {
  const displayName = [ctx.from.first_name, ctx.from.last_name].filter(Boolean).join(' ') || ctx.from.username;
  ctx.flowcastle.identify({ displayName });
  await ctx.reply(
    `Welcome, ${ctx.from.first_name}! This is the FlowCastle Telegraf example.\n\n` +
      'Send any text and I will echo it back, or tap the button to record a demo goal.',
    Markup.inlineKeyboard([Markup.button.callback('🎯 Record demo goal', 'demo_goal')]),
  );
});

bot.help((ctx) =>
  ctx.reply(
    '/start — greet you and identify you as a FlowCastle contact\n' +
      '/qualify — three-step lead qualification, fully in code\n' +
      '/human — hand this chat to a person in FlowCastle Live Chat\n' +
      '/help — this message',
  ),
);

bot.command('qualify', async (ctx) => {
  qualification.set(ctx.from.id, {});
  await ctx.reply(QUALIFICATION_QUESTIONS.need, keyboardFor('need'));
});

bot.command('human', async (ctx) => {
  ctx.flowcastle.requestLiveAgent({ note: 'User asked to talk to a human.' });
  await ctx.reply('Connecting you with a teammate — they will reply right here. 💬');
});

bot.action('demo_goal', async (ctx) => {
  ctx.flowcastle.goal('demo_goal', { source: 'example' });
  await ctx.answerCbQuery('Goal recorded!');
  await ctx.reply('Recorded a demo_goal — check Analytics in your FlowCastle dashboard.');
});

bot.action(/^lead:/, async (ctx) => {
  const answer = parseQualificationCallback(ctx.match.input);
  if (answer === undefined) return ctx.answerCbQuery();

  const current = qualification.get(ctx.from.id);
  const updated = current === undefined ? undefined : saveQualificationAnswer(current, answer.step, answer.value);
  if (updated === undefined) return ctx.answerCbQuery('This step expired. Send /qualify to restart.');

  qualification.set(ctx.from.id, updated);
  await ctx.answerCbQuery('Saved');

  const nextStep = nextQualificationStep(updated);
  if (nextStep !== undefined) return ctx.reply(QUALIFICATION_QUESTIONS[nextStep], keyboardFor(nextStep));

  const result = completedQualification(updated);
  if (result === undefined) return;
  const lead = { leadNeed: result.need, leadBudget: result.budget, leadTimeline: result.timeline };
  qualification.delete(ctx.from.id);

  // Answers become contact traits (visible in the CRM) and a goal (visible in funnels).
  ctx.flowcastle.identify(lead);
  ctx.flowcastle.goal('lead_qualified', lead);
  await ctx.reply(
    'Thanks — your answers are saved.\n\n' +
      `Need: ${qualificationLabel('need', lead.leadNeed)}\n` +
      `Budget: ${qualificationLabel('budget', lead.leadBudget)}\n` +
      `Timeline: ${qualificationLabel('timeline', lead.leadTimeline)}`,
  );

  // Optional: hand off to a follow-up flow that a teammate built in the FlowCastle editor.
  if (FOLLOW_UP_FLOW_KEY !== undefined) {
    try {
      await ctx.flowcastle.runFlow(FOLLOW_UP_FLOW_KEY, { inputs: lead });
    } catch (error) {
      console.error('[flowcastle] follow-up flow failed to start', error);
    }
  }
});

bot.on('text', async (ctx) => {
  // A human is handling this chat in Live Chat — do not echo over them.
  if (ctx.flowcastle.isLiveAgentActive) return;
  await ctx.reply(`Echo: ${ctx.message.text}`);
});

async function shutdown(signal: NodeJS.Signals): Promise<void> {
  console.log(`\n${signal}: shutting down`);
  bot.stop(signal);
  await fc.flush(); // ship anything still buffered
  fc.destroy();
  process.exit(0);
}
process.once('SIGINT', () => void shutdown('SIGINT'));
process.once('SIGTERM', () => void shutdown('SIGTERM'));

async function main(): Promise<void> {
  await fc.ready(); // load the flow manifest before polling
  await bot.launch(() => console.log(`@${bot.botInfo?.username} is up with FlowCastle attached.`));
}
void main().catch((error: unknown) => {
  console.error('Failed to start:', error);
  fc.destroy();
  process.exitCode = 1;
});

File: examples/python-telegram-bot-bot/bot.py
"""A small, realistic python-telegram-bot (v21/22) bot with FlowCastle attached.

Everything below /start is ordinary PTB code. FlowCastle is installed once with
`adapter.install(application)` and then used from handlers via `context.flowcastle` —
identify a contact, record goals, hand a chat to a human, or start a flow that
teammates built visually in the FlowCastle editor.
"""

from __future__ import annotations

import logging
import os
import sys

from dotenv import load_dotenv
from flowcastle import FlowCastleContext, FlowCastleCore, FlowCastleOptions, PrivacyOptions
from flowcastle.adapters.python_telegram_bot import PythonTelegramBotAdapter
from telegram import InlineKeyboardButton, InlineKeyboardMarkup, Update
from telegram.ext import Application, CallbackQueryHandler, CommandHandler, ContextTypes, MessageHandler, filters

from lead_qualification import OPTIONS, QUESTIONS, InMemoryStore, label, parse_callback

load_dotenv()
logging.basicConfig(level=logging.INFO)


def require_env(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        sys.exit(f'Missing required environment variable {name}.')
    return value


BOT_TOKEN = require_env('BOT_TOKEN')
FLOWCASTLE_API_KEY = require_env('FLOWCASTLE_API_KEY')
FOLLOW_UP_FLOW_KEY = os.environ.get('FLOWCASTLE_FOLLOW_UP_FLOW_KEY')

core = FlowCastleCore(FlowCastleOptions(
    api_key=FLOWCASTLE_API_KEY,
    # Routing-only content: commands and callback ids are shared, free text is not.
    privacy=PrivacyOptions(contact_fields=('username', 'language_code')),
    # Lets flows built in the FlowCastle editor run through this bot process.
    runtime_enabled=True,
))
adapter = PythonTelegramBotAdapter(core)
qualification = InMemoryStore()


def fc(context: ContextTypes.DEFAULT_TYPE) -> FlowCastleContext:
    """The FlowCastle context the adapter attaches to every unmatched update."""
    return getattr(context, 'flowcastle')


def keyboard_for(step: str) -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup([[InlineKeyboardButton(text, callback_data=f'lead:{step}:{value}')] for text, value in OPTIONS[step]])


async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    user = update.effective_user
    if update.message is None or user is None:
        return
    await fc(context).identify({'displayName': user.full_name})
    await update.message.reply_text(
        f'Welcome, {user.first_name}! This is the FlowCastle python-telegram-bot example.\n\n'
        'Send any text and I will echo it back, or tap the button to record a demo goal.',
        reply_markup=InlineKeyboardMarkup([[InlineKeyboardButton('🎯 Record demo goal', callback_data='demo_goal')]]),
    )


async def help_command(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if update.message is None:
        return
    await update.message.reply_text(
        '/start — greet you and identify you as a FlowCastle contact\n'
        '/qualify — three-step lead qualification, fully in code\n'
        '/human — hand this chat to a person in FlowCastle Live Chat\n'
        '/help — this message'
    )


async def qualify(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if update.message is None or update.effective_user is None:
        return
    qualification.start(update.effective_user.id)
    await update.message.reply_text(QUESTIONS['need'], reply_markup=keyboard_for('need'))


async def human(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if update.message is None:
        return
    await fc(context).request_live_agent('User asked to talk to a human.')
    await update.message.reply_text('Connecting you with a teammate — they will reply right here. 💬')


async def demo_goal(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    query = update.callback_query
    if query is None:
        return
    await fc(context).goal('demo_goal', {'source': 'example'})
    await query.answer('Goal recorded!')
    if update.effective_chat is not None:
        await context.bot.send_message(update.effective_chat.id, 'Recorded a demo_goal — check Analytics in your FlowCastle dashboard.')


async def qualification_answer(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    query = update.callback_query
    if query is None or update.effective_user is None or update.effective_chat is None:
        return
    parsed = parse_callback(query.data or '')
    state = qualification.get(update.effective_user.id)
    if parsed is None or state is None or not state.save(*parsed):
        await query.answer('This step expired. Send /qualify to restart.')
        return
    await query.answer('Saved')
    chat_id = update.effective_chat.id

    next_step = state.next_step()
    if next_step is not None:
        await context.bot.send_message(chat_id, QUESTIONS[next_step], reply_markup=keyboard_for(next_step))
        return

    lead = state.completed()
    if lead is None:
        return
    qualification.delete(update.effective_user.id)

    # Answers become contact traits (visible in the CRM) and a goal (visible in funnels).
    await fc(context).identify(lead)
    await fc(context).goal('lead_qualified', lead)
    await context.bot.send_message(
        chat_id,
        'Thanks — your answers are saved.\n\n'
        f"Need: {label('need', lead['leadNeed'])}\n"
        f"Budget: {label('budget', lead['leadBudget'])}\n"
        f"Timeline: {label('timeline', lead['leadTimeline'])}",
    )

    # Optional: hand off to a follow-up flow that a teammate built in the FlowCastle editor.
    if FOLLOW_UP_FLOW_KEY:
        try:
            await fc(context).run_flow(FOLLOW_UP_FLOW_KEY, lead)
        except Exception:  # noqa: BLE001 — never let FlowCastle break the bot
            logging.exception('[flowcastle] follow-up flow failed to start')


async def echo(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if update.message is not None and update.message.text is not None:
        await update.message.reply_text(f'Echo: {update.message.text}')


async def on_startup(application: Application) -> None:
    await adapter.ready()  # load the flow manifest before polling


async def on_shutdown(application: Application) -> None:
    await adapter.stop()  # final bounded flush


def main() -> None:
    application = Application.builder().token(BOT_TOKEN).post_init(on_startup).post_shutdown(on_shutdown).build()
    adapter.install(application)  # early handler group; matched updates never reach the handlers below

    application.add_handler(CommandHandler('start', start))
    application.add_handler(CommandHandler('help', help_command))
    application.add_handler(CommandHandler('qualify', qualify))
    application.add_handler(CommandHandler('human', human))
    application.add_handler(CallbackQueryHandler(demo_goal, pattern='^demo_goal$'))
    application.add_handler(CallbackQueryHandler(qualification_answer, pattern='^lead:'))
    application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, echo))

    logging.info('python-telegram-bot is up with FlowCastle attached.')
    application.run_polling()


if __name__ == '__main__':
    main()
